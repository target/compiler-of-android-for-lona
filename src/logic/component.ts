import * as XML from '../xml/ast'
import { FunctionDeclaration } from './nodes/FunctionDeclaration'
import {
  createView,
  ViewOptions,
  createTextView,
  TextViewOptions,
} from '../android/layoutResources'
import { AST } from '@lona/compiler/lib/helpers/logic-ast'
import { findNode } from './syntaxNode'
import { FunctionCallExpression } from './nodes/FunctionCallExpression'
import { IdentifierExpression } from './nodes/IdentifierExpression'
import { IExpression } from './nodes/interfaces'
import { EvaluationContext } from './evaluation'
import { Value } from './runtime/value'
import { LiteralExpression } from './nodes/LiteralExpression'
import { ArrayLiteral } from './nodes/literals'

class ComponentVisitor {
  evaluation: EvaluationContext
  intrinsicNameCount: { [key: string]: number } = {}

  constructor(evaluation: EvaluationContext) {
    this.evaluation = evaluation
  }

  private createIntrinsicNameCount(type: string): number {
    const count = this.intrinsicNameCount[type] || 1

    this.intrinsicNameCount[type] = count + 1

    return count
  }

  createIntrinsicName(type: string): string {
    const count = this.createIntrinsicNameCount(type)

    return count === 1 ? type.toLowerCase() : `${type.toLowerCase()}${count}`
  }
}

type DimensionSize =
  | {
      type: 'fixed'
      value: number
    }
  | { type: 'flexible' }

function getString({ type, memory }: Value): string | undefined {
  if (
    type.type === 'constant' &&
    type.name === 'String' &&
    memory.type === 'string'
  ) {
    return memory.value
  }
}

function getColor({ type, memory }: Value): string | undefined {
  if (
    type.type === 'constant' &&
    type.name === 'Color' &&
    memory.type === 'record'
  ) {
    const colorValue = memory.value['value']
    return getString(colorValue)
  }
}

function getDimensionSize({ type, memory }: Value): DimensionSize | undefined {
  if (
    type.type === 'constant' &&
    type.name === 'DimensionSize' &&
    memory.type === 'enum'
  ) {
    const { value, data } = memory

    switch (value) {
      case 'fixed':
        if (data.length !== 1 || data[0].memory.type !== 'number') {
          throw new Error(`Bad DimensionSize data: ${JSON.stringify(data)}`)
        }

        return { type: value, value: data[0].memory.value }
      case 'flexible':
        return { type: value }
      default:
        throw new Error(`Bad DimensionSize case: ${value}`)
    }
  }
}

function convertDimensionSize(
  dimensionSize: DimensionSize
): string | undefined {
  switch (dimensionSize.type) {
    case 'fixed':
      return `${dimensionSize.value}dp`
    case 'flexible':
      return 'match_parent'
    default:
      break
  }
}

type AndroidView =
  | {
      id: string
      type: 'View'
      options: ViewOptions
      children: AndroidView[]
    }
  | {
      id: string
      type: 'TextView'
      options: TextViewOptions
    }

const createAndroidId = (string: string) => `@+id/${string}`
const createAndroidIdReference = (string: string) => `@id/${string}`

function createViewHierarchy(
  visitor: ComponentVisitor,
  node: IExpression
): AndroidView {
  if (node instanceof FunctionCallExpression) {
    const { callee, argumentExpressionNodes } = node

    if (!(callee instanceof IdentifierExpression)) {
      throw new Error('Element function must be an identifier (for now)')
    }

    function getViewName(): string | undefined {
      if ('__name' in argumentExpressionNodes) {
        const expression = argumentExpressionNodes['__name']
        const value = visitor.evaluation.evaluate(expression.id)
        const id = value && getString(value)
        return id
      }
    }

    const id =
      getViewName()?.toLowerCase() || visitor.createIntrinsicName(callee.name)

    let viewOptions: ViewOptions = {
      id: createAndroidId(id),
    }

    let children: AndroidView[] = []

    if ('width' in argumentExpressionNodes) {
      const expression = argumentExpressionNodes['width']
      const value = visitor.evaluation.evaluate(expression.id)
      const dimensionSize = value && getDimensionSize(value)
      viewOptions.layoutWidth =
        dimensionSize && convertDimensionSize(dimensionSize)
    }

    if ('height' in argumentExpressionNodes) {
      const expression = argumentExpressionNodes['height']
      const value = visitor.evaluation.evaluate(expression.id)
      const dimensionSize = value && getDimensionSize(value)
      viewOptions.layoutHeight =
        dimensionSize && convertDimensionSize(dimensionSize)
    }

    if ('backgroundColor' in argumentExpressionNodes) {
      const expression = argumentExpressionNodes['backgroundColor']
      const value = visitor.evaluation.evaluate(expression.id)
      const background = value && getColor(value)
      viewOptions.background = background
    }

    if ('children' in argumentExpressionNodes) {
      const expression = argumentExpressionNodes['children']

      const literal =
        expression instanceof LiteralExpression && expression.literal

      if (literal instanceof ArrayLiteral) {
        children = literal.elements.map(expression =>
          createViewHierarchy(visitor, expression)
        )
      }
    }

    switch (callee.name) {
      case 'View': {
        return { id: id, type: 'View', options: viewOptions, children }
      }
      case 'HorizontalStack': {
        // Start constraint
        children.forEach((child, index) => {
          if (index === 0) {
            child.options.constraintStartToStartOf = 'parent'
          } else {
            child.options.constraintStartToEndOf = createAndroidIdReference(
              children[index - 1].id
            )
          }
        })

        // End constraint
        children.forEach((child, index, list) => {
          if (index < list.length - 1) {
            child.options.constraintEndToStartOf = createAndroidIdReference(
              children[index + 1].id
            )
          } else {
            child.options.constraintEndToEndOf = 'parent'
          }
        })

        // Secondary axis constraint
        children.forEach(child => {
          child.options.constraintTopToTopOf = 'parent'

          if (!child.options.layoutHeight) {
            child.options.constraintBottomToBottomOf = 'parent'
          }
        })

        return { id: id, type: 'View', options: viewOptions, children }
      }
      case 'Text': {
        const textViewOptions: TextViewOptions = { ...viewOptions }

        if ('value' in argumentExpressionNodes) {
          const expression = argumentExpressionNodes['value']
          const value = visitor.evaluation.evaluate(expression.id)
          const text = value && getString(value)
          if (text) {
            textViewOptions.text = text
          }
        }

        return { id: id, type: 'TextView', options: textViewOptions }
      }
      default:
        throw new Error(`Unknown element name: ${callee.name}`)
    }
  }

  throw new Error('Unhandled element')
}

function createElementTree(view: AndroidView): XML.Element {
  switch (view.type) {
    case 'View': {
      const { options, children } = view
      return createView(options, children.map(createElementTree))
    }
    case 'TextView': {
      const { options } = view
      return createTextView(options)
    }
  }
}

export function createLayout(
  { evaluationContext }: { evaluationContext: EvaluationContext },
  node: FunctionDeclaration
): XML.Element {
  const returnStatements = node.returnStatements

  if (returnStatements.length !== 1) {
    throw new Error('Expected a single return from component function')
  }

  const returnExpression = returnStatements[0].expression

  const visitor = new ComponentVisitor(evaluationContext)

  const viewHierarchy = createViewHierarchy(visitor, returnExpression)

  return createElementTree(viewHierarchy)
}

export function findComponentFunction(
  node: AST.SyntaxNode
): FunctionDeclaration | undefined {
  const syntaxNode = findNode(node, child => child.type === 'function') as
    | AST.FunctionDeclaration
    | undefined

  if (syntaxNode) {
    const node = new FunctionDeclaration(syntaxNode)
    return node
  }
}
