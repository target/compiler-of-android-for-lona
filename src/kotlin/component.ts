import { createFs, IFS } from 'buffs'
import upperFirst from 'lodash.upperfirst'
import path from 'path'
import { formatDrawableName } from '../android/drawableResources'
import { createLayoutFile } from '../android/layoutResources'
import { createComponentClass } from '../kotlin/componentClass'
import {
  ComponentParameter,
  convertComponentParameter,
} from '../kotlin/componentParameter'
import { createLayout, findComponentFunction } from '../android/component'
import { ModuleContext } from '../logic/module'
import * as XML from '../xml/ast'
import {
  createAttr,
  createStyleableDeclaration,
} from '../android/valueResources'
import { compact } from '../utils/sequence'
import snakeCase from 'lodash.snakecase'

export function convertComponentFiles(
  moduleContext: ModuleContext,
  packageName: string
): { layoutResources: IFS; componentFiles: IFS; attrResources: XML.Element[] } {
  const { componentFiles, evaluationContext } = moduleContext

  const { fs: layoutResources } = createFs()
  const { fs: components } = createFs()
  const attrResources: XML.Element[] = []

  componentFiles.forEach(({ sourcePath, rootNode }) => {
    const componentFunction = findComponentFunction(rootNode)

    if (!componentFunction) return

    const parameters = componentFunction.parameters

    const componentParameters: ComponentParameter[] = parameters.map(
      parameter => convertComponentParameter(moduleContext, parameter)
    )

    const basename = path.basename(sourcePath)
    const componentName = upperFirst(
      path.basename(sourcePath, path.extname(sourcePath))
    )
    const className = `${componentName}View`

    const attrs: XML.Element[] = compact(
      componentParameters.map(parameter => {
        if (parameter.styleableAttribute) {
          return createAttr(
            snakeCase(parameter.name),
            parameter.styleableAttribute.format
          )
        }
      })
    )

    if (attrs.length > 0) {
      attrResources.push(createStyleableDeclaration(className, attrs))
    }

    const { layout, imports, publicViews } = createLayout(
      { evaluationContext },
      componentFunction
    )

    const componentClass = createComponentClass({
      namePrefix: componentName,
      packagePath: packageName,
      imports,
      parameters: componentParameters,
      publicViews,
    })
    const kotlinFileName = `${className}.kt`
    components.writeFileSync(`/${kotlinFileName}`, componentClass)

    const layoutFile = createLayoutFile(layout)
    const xmlFileName = formatDrawableName(basename, 'xml', {
      nameTemplate: '${qualifiedName?join("-")}',
    })
    layoutResources.writeFileSync(`/${xmlFileName}`, layoutFile)
  })

  return { layoutResources, componentFiles: components, attrResources }
}
