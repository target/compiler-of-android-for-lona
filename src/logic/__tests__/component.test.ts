import { AST } from '@lona/compiler/lib/helpers/logic-ast'
import * as Serialization from '@lona/serialization'
import { createLayout, findComponentFunction } from '../component'
import { run } from '../environment'
import { EvaluationContext } from '../evaluation'
import {
  createLayoutFile,
  createConstraintLayout,
} from '../../android/layoutResources'

function standardEvaluate(rootNode: AST.SyntaxNode): EvaluationContext {
  const evaluation = run(console, [rootNode])

  if (!evaluation) {
    throw new Error('Failed to evaluate')
  }

  return evaluation
}

it('creates View layout', () => {
  const file = `
let primary: Color = #color(css: "#FAB")

func Row() -> Element {
  return View(__name: "Container", backgroundColor: primary, width: DimensionSize.fixed(100), height: DimensionSize.fixed(100))
}
`
  const rootNode = Serialization.decodeLogic(file)
  const evaluationContext = standardEvaluate(rootNode)
  const componentFunction = findComponentFunction(rootNode)

  expect(componentFunction).toBeDefined()
  expect(evaluationContext).toBeDefined()

  if (!componentFunction || !evaluationContext) return

  const { layout } = createLayout({ evaluationContext }, componentFunction)
  const layoutFile = createLayoutFile(layout)

  expect(layoutFile).toMatchSnapshot()
})

it('creates View with children layout', () => {
  const file = `
let primary: Color = #color(css: "#FAB")

func Row() -> Element {
  return View(__name: "Container", children: [
    View(__name: "Child1"),
    View(__name: "Child2")
  ])
}
`
  const rootNode = Serialization.decodeLogic(file)
  const evaluationContext = standardEvaluate(rootNode)
  const componentFunction = findComponentFunction(rootNode)

  expect(componentFunction).toBeDefined()
  expect(evaluationContext).toBeDefined()

  if (!componentFunction || !evaluationContext) return

  const { layout } = createLayout({ evaluationContext }, componentFunction)
  const layoutFile = createLayoutFile(layout)

  expect(layoutFile).toMatchSnapshot()
})

it('creates Text layout', () => {
  const file = `
let primary: Color = #color(css: "#FAB")

func Row() -> Element {
  return View(children: [
    Text(value: "Hello")
  ])
}
`
  const rootNode = Serialization.decodeLogic(file)
  const evaluationContext = standardEvaluate(rootNode)
  const componentFunction = findComponentFunction(rootNode)

  expect(componentFunction).toBeDefined()
  expect(evaluationContext).toBeDefined()

  if (!componentFunction || !evaluationContext) return

  const { layout } = createLayout({ evaluationContext }, componentFunction)
  const layoutFile = createLayoutFile(layout)

  expect(layoutFile).toMatchSnapshot()
})

it('creates HorizontalStack layout', () => {
  const file = `
let primary: Color = #color(css: "#FAB")

func Row() -> Element {
  return HorizontalStack(__name: "Container", children: [
    View(),
    View(),
    View()
  ])
}
`
  const rootNode = Serialization.decodeLogic(file)
  const evaluationContext = standardEvaluate(rootNode)
  const componentFunction = findComponentFunction(rootNode)

  expect(componentFunction).toBeDefined()
  expect(evaluationContext).toBeDefined()

  if (!componentFunction || !evaluationContext) return

  const { layout } = createLayout({ evaluationContext }, componentFunction)
  const layoutFile = createLayoutFile(layout)

  expect(layoutFile).toMatchSnapshot()
})
