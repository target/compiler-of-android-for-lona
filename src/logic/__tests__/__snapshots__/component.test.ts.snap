// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`creates HorizontalStack layout 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<androidx.constraintlayout.widget.ConstraintLayout
    xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
    xmlns:app=\\"http://schemas.android.com/apk/res-auto\\"
    xmlns:tools=\\"http://schemas.android.com/tools\\"
    android:layout_width=\\"match_parent\\"
    android:layout_height=\\"match_parent\\">
  <View
      android:id=\\"@+id/container\\"
      app:layout_constraintTop_toTopOf=\\"parent\\"
      app:layout_constraintLeft_toLeftOf=\\"parent\\">
    <View
        android:id=\\"@+id/view\\"
        app:layout_constraintStart_toStartOf=\\"parent\\"
        app:layout_constraintEnd_toStartOf=\\"@id/view2\\"
        app:layout_constraintTop_toTopOf=\\"parent\\"
        app:layout_constraintBottom_toBottomOf=\\"parent\\"
        app:layout_constraintTop_toTopOf=\\"parent\\"
        app:layout_constraintLeft_toLeftOf=\\"parent\\" />
    <View
        android:id=\\"@+id/view2\\"
        app:layout_constraintStart_toEndOf=\\"@id/view\\"
        app:layout_constraintEnd_toStartOf=\\"@id/view3\\"
        app:layout_constraintTop_toTopOf=\\"parent\\"
        app:layout_constraintBottom_toBottomOf=\\"parent\\"
        app:layout_constraintTop_toTopOf=\\"parent\\"
        app:layout_constraintLeft_toLeftOf=\\"parent\\" />
    <View
        android:id=\\"@+id/view3\\"
        app:layout_constraintStart_toEndOf=\\"@id/view2\\"
        app:layout_constraintEnd_toEndOf=\\"parent\\"
        app:layout_constraintTop_toTopOf=\\"parent\\"
        app:layout_constraintBottom_toBottomOf=\\"parent\\"
        app:layout_constraintTop_toTopOf=\\"parent\\"
        app:layout_constraintLeft_toLeftOf=\\"parent\\" />
  </View>
</androidx.constraintlayout.widget.ConstraintLayout>"
`;

exports[`creates Text layout 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<androidx.constraintlayout.widget.ConstraintLayout
    xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
    xmlns:app=\\"http://schemas.android.com/apk/res-auto\\"
    xmlns:tools=\\"http://schemas.android.com/tools\\"
    android:layout_width=\\"match_parent\\"
    android:layout_height=\\"match_parent\\">
  <TextView
      android:id=\\"@+id/container\\"
      app:layout_constraintTop_toTopOf=\\"parent\\"
      app:layout_constraintLeft_toLeftOf=\\"parent\\"
      android:text=\\"Hello\\" />
</androidx.constraintlayout.widget.ConstraintLayout>"
`;

exports[`creates View layout 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<androidx.constraintlayout.widget.ConstraintLayout
    xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
    xmlns:app=\\"http://schemas.android.com/apk/res-auto\\"
    xmlns:tools=\\"http://schemas.android.com/tools\\"
    android:layout_width=\\"match_parent\\"
    android:layout_height=\\"match_parent\\">
  <View
      android:id=\\"@+id/container\\"
      android:layout_width=\\"100dp\\"
      android:layout_height=\\"100dp\\"
      android:background=\\"#FAB\\"
      app:layout_constraintTop_toTopOf=\\"parent\\"
      app:layout_constraintLeft_toLeftOf=\\"parent\\" />
</androidx.constraintlayout.widget.ConstraintLayout>"
`;

exports[`creates View with children layout 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"utf-8\\"?>
<androidx.constraintlayout.widget.ConstraintLayout
    xmlns:android=\\"http://schemas.android.com/apk/res/android\\"
    xmlns:app=\\"http://schemas.android.com/apk/res-auto\\"
    xmlns:tools=\\"http://schemas.android.com/tools\\"
    android:layout_width=\\"match_parent\\"
    android:layout_height=\\"match_parent\\">
  <View
      android:id=\\"@+id/container\\"
      app:layout_constraintTop_toTopOf=\\"parent\\"
      app:layout_constraintLeft_toLeftOf=\\"parent\\">
    <View
        android:id=\\"@+id/child1\\"
        app:layout_constraintTop_toTopOf=\\"parent\\"
        app:layout_constraintLeft_toLeftOf=\\"parent\\" />
    <View
        android:id=\\"@+id/child2\\"
        app:layout_constraintTop_toTopOf=\\"parent\\"
        app:layout_constraintLeft_toLeftOf=\\"parent\\" />
  </View>
</androidx.constraintlayout.widget.ConstraintLayout>"
`;
